Stats: Make Upsell section recognize Complete plans

a8ck3n opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to:

- Issue: https://github.com/Automattic/wp-calypso/issues/80536
- Project: pejTkB-1zl-p2

## Proposed Changes

Odyssey Stats shows an Upsell section at the bottom of the page. With this change, sites with a Complete plan will no longer see the upsell.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

The upsell doesn't make sense for sites with a Complete plan.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Set up per FG instructions: PCYsg-Pp7-p2
* Check a site without a Complete plan and confirm the upsell is present.
* Check a site with a Complete plan and confirm the upsell is no longer shown.

Note: The request is cached on the server side (I think) so the `purchases` API won't immediately reflect plan changes made through Store Admin. Better to test on two separate sites. You can confirm the page is properly reflecting the API response by inspecting it in the dev tools.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

a8ck3n committed:
Make Upsell section recognize Complete plans

Odyssey Stats shows an Upsell section at the bottom of the page. With this change, sites with a Complete plan will no longer see the upsell.

dognose24 commented on client/my-sites/stats/jetpack-upsell-section/index.tsx:
```diff
@@ -27,11 +28,20 @@ export default function JetpackUpsellSection() {
+	// Currently checking for variations of the Complete bundle only.
+	const siteId = useSelector( getSelectedSiteId );
+	const sitePurchases = useSelector( ( state ) => getSitePurchases( state, siteId ) );
+	const hasCompleteBundle = sitePurchases?.some( ( purchase ) =>
```
Could we reuse the function [filterPurchasesByProducts](https://github.com/Automattic/wp-calypso/blob/729b9a212ffc2ad7c3f8d47d7f01dd74d22fed33/client/my-sites/stats/hooks/use-stats-purchases.tsx#L29-L38), which would also check the expiry status?

<img width="1291" alt="截圖 2024-08-02 下午3 56 06" src="https://github.com/user-attachments/assets/52a12914-da0d-411d-86fc-01683a792f1e">

dognose24 commented on client/my-sites/stats/jetpack-upsell-section/index.tsx:
```diff
@@ -27,11 +28,20 @@ export default function JetpackUpsellSection() {
+	// Check for bundled products.
+	// Currently checking for variations of the Complete bundle only.
+	const siteId = useSelector( getSelectedSiteId );
+	const sitePurchases = useSelector( ( state ) => getSitePurchases( state, siteId ) );
```
It would be great to consolidate this purchase check into `useStatsPurchases` for plan support visibility and reusability. What do you think?

a8ck3n replied:
I'll look at this today. I see that this hook uses the `filterPurchasesByProducts` function noted in your previous comment so two comments with one stone. ;>

a8ck3n replied:
So I've looked at extending the `useStatsPurchases` hook for this but I'm not sure it makes sense here. The naming of the hook makes it unclear that it would provide data on other plans and it is obscuring the actual plan data which is what I'd prefer to work with.

I see two approaches that might make sense:

1. Adding additional flags for `hasCompletePlan`, `hasProfessionalPlan`, `hasSecurityBundle`, etc. as needed OR…
2. Adding a `data` value and returning the underlying results from `getSitePurchases` along with the various flags.

I'd lean towards option 2 but my real preference would be to call `getSitePurchases` and work directly with the returned data.

Is there a reason to prefer extending `useStatsPurchases` over calling the `getSitePurchases` directly?

dognose24 replied:
We don't need to use the hook `useStatsPurchases` here, but importing an independent check function seems an option. 🤔

dognose24 reviewed (commented):
I have tested it, and the expired purchase would also hide the upsell section. On the other hand, how do we handle the legacy Jetpack Business plan for the upsell section?

<img width="1291" alt="截圖 2024-08-02 下午3 56 06" src="https://github.com/user-attachments/assets/9cea5055-c247-4b74-baca-a2fe3ddbf46d">

a8ck3n committed:
Merge branch 'trunk' into update/stats-no-upsell-complete-plan

a8ck3n committed:
Tweak logic for hiding upsell section

- Extracted check into new helper function.
- Checks for legacy Business plans.
- Checks for expired plans.

a8ck3n commented:
> I have tested it, and the expired purchase would also hide the upsell section. On the other hand, how do we handle the legacy Jetpack Business plan for the upsell section?

I think we should probably detect the Business plan and hide the upsell on those sites too. I'll add that check and one for expired plans. Cheers!

dognose24 commented on client/my-sites/stats/jetpack-upsell-section/index.tsx:
```diff
@@ -23,15 +25,40 @@ const QUERY_VALUES = {
+	// Currenty that means Business or Complete plans.
+	const hasFullFeaturedPlan = plans.some(
+		( purchase ) =>
+			purchase.productSlug.includes( 'complete' ) || purchase.productSlug.includes( 'business' )
```
Could we check with the exact product slug rather than the substring?

dognose24 commented on client/my-sites/stats/jetpack-upsell-section/index.tsx:
```diff
@@ -23,15 +25,40 @@ const QUERY_VALUES = {
+	}
+
+	// Filter out expired plans.
+	const plans = purchases.filter( ( purchase ) => purchase.expiryStatus !== 'expired' );
```
Is importing the `filterPurchasesByProducts` or `areProductsOwned` here a better way than creating another similar check?

a8ck3n replied:
I've updated the code to use `areProductsOwned` with the existing `const` strings.

a8ck3n committed:
Merge branch 'trunk' into update/stats-no-upsell-complete-plan

a8ck3n committed:
Add checks for Business & Complete plans

a8ck3n committed:
Add JETPACK_BUSINESS_PLANS const

a8ck3n committed:
Use new plan checks for upsell visibility

dognose24 reviewed (approved):
Works as described! 👍🏼 Thanks for the patience in iterating changes, @a8ck3n.

a8ck3n committed:
Merge branch 'trunk' into update/stats-no-upsell-complete-plan

a8ck3n committed:
Comments++
